// @base-ui-components/react/cjs/select/item-text/SelectItemText.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectItemText = void 0;
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _useForkRef = require("../../utils/useForkRef");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _SelectRootContext = require("../root/SelectRootContext");
var _SelectItemContext = require("../item/SelectItemContext");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const InnerSelectItemText = /*#__PURE__*/React.forwardRef(function InnerSelectItemText(props, forwardedRef) {
  const {
    className,
    render,
    selected,
    selectedItemTextRef,
    indexRef,
    ...otherProps
  } = props;
  const mergedRef = (0, _useForkRef.useForkRef)(forwardedRef);
  const state = React.useMemo(() => ({}), []);
  const ref = React.useCallback(node => {
    if (mergedRef) {
      mergedRef(node);
    }

    // Wait for the DOM indices to be set.
    queueMicrotask(() => {
      if (selected || selectedItemTextRef.current === null && indexRef.current === 0) {
        selectedItemTextRef.current = node;
      }
    });
  }, [mergedRef, selected, selectedItemTextRef, indexRef]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    ref,
    render: render ?? 'div',
    className,
    state,
    extraProps: otherProps
  });
  return renderElement();
});
process.env.NODE_ENV !== "production" ? InnerSelectItemText.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: _propTypes.default.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  /**
   * @ignore
   */
  indexRef: _propTypes.default.shape({
    current: _propTypes.default.number.isRequired
  }).isRequired,
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.func]),
  /**
   * @ignore
   */
  selected: _propTypes.default.bool.isRequired,
  /**
   * @ignore
   */
  selectedItemTextRef: _propTypes.default.shape({
    current: (props, propName) => {
      if (props[propName] == null) {
        return null;
      }
      if (typeof props[propName] !== 'object' || props[propName].nodeType !== 1) {
        return new Error(`Expected prop '${propName}' to be of type Element`);
      }
      return null;
    }
  }).isRequired
} : void 0;
const MemoizedInnerSelectItemText = /*#__PURE__*/React.memo(InnerSelectItemText);

/**
 * A text label of the select item.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Select](https://base-ui.com/react/components/select)
 */
const SelectItemText = exports.SelectItemText = /*#__PURE__*/React.forwardRef(function SelectItemText(props, forwardedRef) {
  const {
    selected,
    indexRef
  } = (0, _SelectItemContext.useSelectItemContext)();
  const {
    selectedItemTextRef
  } = (0, _SelectRootContext.useSelectRootContext)();
  const mergedRef = (0, _useForkRef.useForkRef)(forwardedRef);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(MemoizedInnerSelectItemText, {
    ref: mergedRef,
    selected: selected,
    selectedItemTextRef: selectedItemTextRef,
    indexRef: indexRef,
    ...props
  });
});
process.env.NODE_ENV !== "production" ? SelectItemText.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: _propTypes.default.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.func])
} : void 0;
