// @base-ui-components/react/cjs/number-field/scrub-area-cursor/NumberFieldScrubAreaCursorDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NumberFieldScrubAreaCursorDataAttributes = void 0;
let NumberFieldScrubAreaCursorDataAttributes = exports.NumberFieldScrubAreaCursorDataAttributes = /*#__PURE__*/function (NumberFieldScrubAreaCursorDataAttributes) {
  NumberFieldScrubAreaCursorDataAttributes["scrubbing"] = "data-scrubbing";
  NumberFieldScrubAreaCursorDataAttributes["disabled"] = "data-disabled";
  NumberFieldScrubAreaCursorDataAttributes["readonly"] = "data-readonly";
  NumberFieldScrubAreaCursorDataAttributes["required"] = "data-required";
  NumberFieldScrubAreaCursorDataAttributes["valid"] = "data-valid";
  NumberFieldScrubAreaCursorDataAttributes["invalid"] = "data-invalid";
  NumberFieldScrubAreaCursorDataAttributes["touched"] = "data-touched";
  NumberFieldScrubAreaCursorDataAttributes["dirty"] = "data-dirty";
  NumberFieldScrubAreaCursorDataAttributes["filled"] = "data-filled";
  NumberFieldScrubAreaCursorDataAttributes["focused"] = "data-focused";
  return NumberFieldScrubAreaCursorDataAttributes;
}({});
