// @base-ui-components/react/cjs/composite/item/CompositeItem.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompositeItem = CompositeItem;
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useForkRef = require("../../utils/useForkRef");
var _CompositeRootContext = require("../root/CompositeRootContext");
var _useCompositeItem = require("./useCompositeItem");
var _proptypes = require("../../utils/proptypes");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @ignore - internal component.
 */
function CompositeItem(props) {
  const {
    render,
    className,
    itemRef,
    metadata,
    ...otherProps
  } = props;
  const {
    highlightedIndex
  } = (0, _CompositeRootContext.useCompositeRootContext)();
  const {
    getItemProps,
    ref,
    index
  } = (0, _useCompositeItem.useCompositeItem)({
    metadata
  });
  const state = React.useMemo(() => ({
    highlighted: index === highlightedIndex
  }), [index, highlightedIndex]);
  const mergedRef = (0, _useForkRef.useForkRef)(itemRef, ref);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    propGetter: getItemProps,
    ref: mergedRef,
    render: render ?? 'div',
    state,
    className,
    extraProps: otherProps
  });
  return renderElement();
}
process.env.NODE_ENV !== "production" ? CompositeItem.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: _propTypes.default.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  /**
   * @ignore
   */
  itemRef: _proptypes.refType,
  /**
   * @ignore
   */
  metadata: _propTypes.default.any,
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.func])
} : void 0;
