// @backstage/canon/dist/components/Grid/Grid.props.esm.js
const columnsValues = [
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "11",
  "12",
  "auto"
];
const gridPropDefs = {
  columns: {
    type: "enum | string",
    className: "cu-columns",
    customProperties: ["--columns"],
    values: columnsValues,
    responsive: true,
    default: "auto"
  }
};
const gridItemPropDefs = {
  colSpan: {
    type: "enum | string",
    className: "cu-col-span",
    customProperties: ["--col-span"],
    values: columnsValues,
    responsive: true
  },
  colEnd: {
    type: "enum | string",
    className: "cu-col-end",
    customProperties: ["--col-end"],
    values: columnsValues,
    responsive: true
  },
  colStart: {
    type: "enum | string",
    className: "cu-col-start",
    customProperties: ["--col-start"],
    values: columnsValues,
    responsive: true
  },
  rowSpan: {
    type: "enum | string",
    className: "cu-row-span",
    customProperties: ["--row-span"],
    values: columnsValues,
    responsive: true
  }
};

export { gridItemPropDefs, gridPropDefs };
