// @backstage/canon/dist/props/margin.props.esm.js
const marginPropDefs = (spacingValues) => ({
  m: {
    type: "enum | string",
    values: spacingValues,
    className: "cu-m",
    customProperties: ["--m"],
    responsive: true
  },
  mx: {
    type: "enum | string",
    values: spacingValues,
    className: "cu-mx",
    customProperties: ["--mx"],
    responsive: true
  },
  my: {
    type: "enum | string",
    values: spacingValues,
    className: "cu-my",
    customProperties: ["--my"],
    responsive: true
  },
  mt: {
    type: "enum | string",
    values: spacingValues,
    className: "cu-mt",
    customProperties: ["--mt"],
    responsive: true
  },
  mr: {
    type: "enum | string",
    values: spacingValues,
    className: "cu-mr",
    customProperties: ["--mr"],
    responsive: true
  },
  mb: {
    type: "enum | string",
    values: spacingValues,
    className: "cu-mb",
    customProperties: ["--mb"],
    responsive: true
  },
  ml: {
    type: "enum | string",
    values: spacingValues,
    className: "cu-ml",
    customProperties: ["--ml"],
    responsive: true
  }
});

export { marginPropDefs };
