// @backstage/canon/dist/props/padding.props.esm.js
const paddingPropDefs = (spacingValues) => ({
  p: {
    type: "enum | string",
    className: "cu-p",
    customProperties: ["--p"],
    values: spacingValues,
    responsive: true
  },
  px: {
    type: "enum | string",
    className: "cu-px",
    customProperties: ["--px"],
    values: spacingValues,
    responsive: true
  },
  py: {
    type: "enum | string",
    className: "cu-py",
    customProperties: ["--py"],
    values: spacingValues,
    responsive: true
  },
  pt: {
    type: "enum | string",
    className: "cu-pt",
    customProperties: ["--pt"],
    values: spacingValues,
    responsive: true
  },
  pr: {
    type: "enum | string",
    className: "cu-pr",
    customProperties: ["--pr"],
    values: spacingValues,
    responsive: true
  },
  pb: {
    type: "enum | string",
    className: "cu-pb",
    customProperties: ["--pb"],
    values: spacingValues,
    responsive: true
  },
  pl: {
    type: "enum | string",
    className: "cu-pl",
    customProperties: ["--pl"],
    values: spacingValues,
    responsive: true
  }
});

export { paddingPropDefs };
