// @backstage/plugin-kubernetes-react/dist/components/StatefulSetsAccordions/StatefulSetDrawer.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { KubernetesStructuredMetadataTableDrawer } from '../KubernetesDrawer/KubernetesStructuredMetadataTableDrawer.esm.js';
import '../KubernetesDrawer/KubernetesDrawer.esm.js';
import '@backstage/core-components';
import '@material-ui/core/FormControlLabel';
import '@material-ui/core/Switch';
import 'js-yaml';
import 'react';
import { renderCondition } from '../../utils/pod.esm.js';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import Chip from '@material-ui/core/Chip';

const StatefulSetDrawer = ({
  statefulset,
  expanded
}) => {
  const namespace = statefulset.metadata?.namespace;
  return /* @__PURE__ */ jsx(
    KubernetesStructuredMetadataTableDrawer,
    {
      object: statefulset,
      expanded,
      kind: "StatefulSet",
      renderObject: (statefulsetObj) => {
        const conditions = (statefulsetObj.status?.conditions ?? []).map(renderCondition).reduce((accum, next) => {
          accum[next[0]] = next[1];
          return accum;
        }, {});
        return {
          updateStrategy: statefulset.spec?.updateStrategy ?? "???",
          podManagementPolicy: statefulset.spec?.podManagementPolicy ?? "???",
          serviceName: statefulset.spec?.serviceName ?? "???",
          selector: statefulset.spec?.selector ?? "???",
          revisionHistoryLimit: statefulset.spec?.revisionHistoryLimit ?? "???",
          ...conditions
        };
      },
      children: /* @__PURE__ */ jsxs(
        Grid,
        {
          container: true,
          direction: "column",
          justifyContent: "flex-start",
          alignItems: "flex-start",
          spacing: 0,
          children: [
            /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Typography, { variant: "body1", children: statefulset.metadata?.name ?? "unknown object" }) }),
            /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Typography, { color: "textSecondary", variant: "subtitle1", children: "Stateful Set" }) }),
            namespace && /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Chip, { size: "small", label: `namespace: ${namespace}` }) })
          ]
        }
      )
    }
  );
};

export { StatefulSetDrawer };
