// @base-ui-components/react/esm/dialog/description/DialogDescription.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useDialogRootContext } from '../root/DialogRootContext.js';
import { mergeProps } from '../../merge-props/index.js';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
import { useEnhancedEffect } from '../../utils/useEnhancedEffect.js';
import { useBaseUiId } from '../../utils/useBaseUiId.js';
const state = {};

/**
 * A paragraph with additional information about the dialog.
 * Renders a `<p>` element.
 *
 * Documentation: [Base UI Dialog](https://base-ui.com/react/components/dialog)
 */
const DialogDescription = /*#__PURE__*/React.forwardRef(function DialogDescription(props, forwardedRef) {
  const {
    render,
    className,
    id: idProp,
    ...other
  } = props;
  const {
    setDescriptionElementId
  } = useDialogRootContext();
  const id = useBaseUiId(idProp);
  useEnhancedEffect(() => {
    setDescriptionElementId(id);
    return () => {
      setDescriptionElementId(undefined);
    };
  }, [id, setDescriptionElementId]);
  const getProps = React.useCallback((externalProps = {}) => mergeProps({
    id
  }, externalProps), [id]);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getProps,
    render: render ?? 'p',
    className,
    state,
    ref: forwardedRef,
    extraProps: other
  });
  return renderElement();
});
process.env.NODE_ENV !== "production" ? DialogDescription.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * @ignore
   */
  id: PropTypes.string,
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
export { DialogDescription };
