// @backstage/canon/dist/components/Tabs/Tabs.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { Tabs as Tabs$1 } from '@base-ui-components/react/tabs';
import clsx from 'clsx';

const TabsRoot = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Tabs$1.Root,
  {
    ref,
    className: clsx("canon-TabsRoot", className),
    ...props
  }
));
TabsRoot.displayName = Tabs$1.Root.displayName;
const TabsList = forwardRef(({ className, children, ...props }, ref) => /* @__PURE__ */ jsxs(
  Tabs$1.List,
  {
    ref,
    className: clsx("canon-TabsList", className),
    ...props,
    children: [
      children,
      /* @__PURE__ */ jsx(Tabs$1.Indicator, { className: "canon-TabsIndicator" })
    ]
  }
));
TabsList.displayName = Tabs$1.List.displayName;
const TabsTab = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Tabs$1.Tab,
  {
    ref,
    className: clsx("canon-TabsTab", className),
    ...props
  }
));
TabsTab.displayName = Tabs$1.Tab.displayName;
const TabsPanel = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Tabs$1.Panel,
  {
    ref,
    className: clsx("canon-TabsPanel", className),
    ...props
  }
));
TabsPanel.displayName = Tabs$1.Panel.displayName;
const Tabs = {
  Root: TabsRoot,
  List: TabsList,
  Tab: TabsTab,
  Panel: TabsPanel
};

export { Tabs };
