// @base-ui-components/react/esm/slider/value/SliderValue.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
import { useSliderRootContext } from '../root/SliderRootContext.js';
import { sliderStyleHookMapping } from '../root/styleHooks.js';
import { useSliderValue } from './useSliderValue.js';
/**
 * Displays the current value of the slider as text.
 * Renders an `<output>` element.
 *
 * Documentation: [Base UI Slider](https://base-ui.com/react/components/slider)
 */
const SliderValue = /*#__PURE__*/React.forwardRef(function SliderValue(props, forwardedRef) {
  const {
    'aria-live': ariaLive = 'off',
    render,
    className,
    children,
    ...otherProps
  } = props;
  const {
    thumbMap,
    state,
    values,
    format
  } = useSliderRootContext();
  const {
    getRootProps,
    formattedValues
  } = useSliderValue({
    'aria-live': ariaLive,
    format: format ?? null,
    thumbMap,
    values
  });
  const defaultDisplayValue = React.useMemo(() => {
    const arr = [];
    for (let i = 0; i < values.length; i += 1) {
      arr.push(formattedValues[i] || values[i]);
    }
    return arr.join(' – ');
  }, [values, formattedValues]);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getRootProps,
    render: render ?? 'output',
    state,
    className,
    ref: forwardedRef,
    extraProps: {
      children: typeof children === 'function' ? children(formattedValues, values) : defaultDisplayValue,
      ...otherProps
    },
    customStyleHookMapping: sliderStyleHookMapping
  });
  return renderElement();
});
export { SliderValue };
process.env.NODE_ENV !== "production" ? SliderValue.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @default 'off'
   */
  'aria-live': PropTypes.oneOf(['assertive', 'off', 'polite']),
  /**
   * @ignore
   */
  children: PropTypes.func,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
