// @base-ui-components/react/esm/slider/control/SliderControl.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
import { useSliderRootContext } from '../root/SliderRootContext.js';
import { sliderStyleHookMapping } from '../root/styleHooks.js';
import { useSliderControl } from './useSliderControl.js';

/**
 * The clickable, interactive part of the slider.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Slider](https://base-ui.com/react/components/slider)
 */
const SliderControl = /*#__PURE__*/React.forwardRef(function SliderControl(props, forwardedRef) {
  const {
    render: renderProp,
    className,
    ...otherProps
  } = props;
  const {
    commitValue,
    disabled,
    dragging,
    getFingerState,
    lastChangedValueRef,
    minStepsBetweenValues,
    percentageValues,
    registerSliderControl,
    setActive,
    setDragging,
    setValue,
    state,
    step,
    thumbRefs
  } = useSliderRootContext();
  const {
    getRootProps
  } = useSliderControl({
    commitValue,
    disabled,
    dragging,
    getFingerState,
    lastChangedValueRef,
    minStepsBetweenValues,
    percentageValues,
    registerSliderControl,
    rootRef: forwardedRef,
    setActive,
    setDragging,
    setValue,
    step,
    thumbRefs
  });
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getRootProps,
    render: renderProp ?? 'div',
    state,
    className,
    extraProps: otherProps,
    customStyleHookMapping: sliderStyleHookMapping
  });
  return renderElement();
});
export { SliderControl };
process.env.NODE_ENV !== "production" ? SliderControl.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
