// @base-ui-components/react/esm/input/Input.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { Field } from '../field/index.js';

/**
 * A native input element that automatically works with [Field](https://base-ui.com/react/components/field).
 * Renders an `<input>` element.
 *
 * Documentation: [Base UI Input](https://base-ui.com/react/components/input)
 */
import { jsx as _jsx } from "react/jsx-runtime";
const Input = /*#__PURE__*/React.forwardRef(function Input(props, forwardedRef) {
  const {
    render,
    className,
    ...otherProps
  } = props;
  return /*#__PURE__*/_jsx(Field.Control, {
    ref: forwardedRef,
    render: render,
    className: className,
    ...otherProps
  });
});
process.env.NODE_ENV !== "production" ? Input.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
export { Input };
