// @backstage/canon/dist/components/Menu/Menu.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { Menu as Menu$1 } from '@base-ui-components/react/menu';
import clsx from 'clsx';
import { Combobox } from './Combobox.esm.js';
import '@remixicon/react';
import { Icon } from '../Icon/Icon.esm.js';
import '../Icon/context.esm.js';

const MenuTrigger = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.Trigger,
  {
    ref,
    className: clsx("canon-MenuTrigger", className),
    ...props
  }
));
MenuTrigger.displayName = Menu$1.Trigger.displayName;
const MenuBackdrop = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.Backdrop,
  {
    ref,
    className: clsx("canon-MenuBackdrop", className),
    ...props
  }
));
MenuBackdrop.displayName = Menu$1.Backdrop.displayName;
const MenuPositioner = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.Positioner,
  {
    ref,
    className: clsx("canon-MenuPositioner", className),
    ...props
  }
));
MenuPositioner.displayName = Menu$1.Positioner.displayName;
const MenuPopup = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.Popup,
  {
    ref,
    className: clsx("canon-MenuPopup", className),
    ...props
  }
));
MenuPopup.displayName = Menu$1.Popup.displayName;
const MenuArrow = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.Arrow,
  {
    ref,
    className: clsx("canon-MenuArrow", className),
    ...props
  }
));
MenuArrow.displayName = Menu$1.Arrow.displayName;
const MenuItem = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.Item,
  {
    ref,
    className: clsx("canon-MenuItem", className),
    ...props
  }
));
MenuItem.displayName = Menu$1.Item.displayName;
const MenuGroup = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.Group,
  {
    ref,
    className: clsx("canon-MenuGroup", className),
    ...props
  }
));
MenuGroup.displayName = Menu$1.Group.displayName;
const MenuGroupLabel = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.GroupLabel,
  {
    ref,
    className: clsx("canon-MenuGroupLabel", className),
    ...props
  }
));
MenuGroupLabel.displayName = Menu$1.GroupLabel.displayName;
const MenuRadioGroup = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.RadioGroup,
  {
    ref,
    className: clsx("canon-MenuRadioGroup", className),
    ...props
  }
));
MenuRadioGroup.displayName = Menu$1.RadioGroup.displayName;
const MenuRadioItem = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.RadioItem,
  {
    ref,
    className: clsx("canon-MenuRadioItem", className),
    ...props
  }
));
MenuRadioItem.displayName = Menu$1.RadioItem.displayName;
const MenuRadioItemIndicator = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.RadioItemIndicator,
  {
    ref,
    className: clsx("canon-MenuRadioItemIndicator", className),
    ...props
  }
));
MenuRadioItemIndicator.displayName = Menu$1.RadioItemIndicator.displayName;
const MenuCheckboxItem = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.CheckboxItem,
  {
    ref,
    className: clsx("canon-MenuCheckboxItem", className),
    ...props
  }
));
MenuCheckboxItem.displayName = Menu$1.CheckboxItem.displayName;
const MenuCheckboxItemIndicator = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.CheckboxItemIndicator,
  {
    ref,
    className: clsx("canon-MenuCheckboxItemIndicator", className),
    ...props
  }
));
MenuCheckboxItemIndicator.displayName = Menu$1.CheckboxItemIndicator.displayName;
const MenuSubmenuTrigger = forwardRef(({ className, children, ...props }, ref) => /* @__PURE__ */ jsxs(
  Menu$1.SubmenuTrigger,
  {
    ref,
    className: clsx("canon-MenuSubmenuTrigger", className),
    ...props,
    children: [
      /* @__PURE__ */ jsx("div", { children }),
      /* @__PURE__ */ jsx(Icon, { "aria-label": "Submenu indicator icon", name: "chevron-right", size: 20 })
    ]
  }
));
MenuSubmenuTrigger.displayName = Menu$1.SubmenuTrigger.displayName;
const MenuSeparator = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  Menu$1.Separator,
  {
    ref,
    className: clsx("canon-MenuSeparator", className),
    ...props
  }
));
MenuSeparator.displayName = Menu$1.Separator.displayName;
const Menu = {
  Root: Menu$1.Root,
  Trigger: MenuTrigger,
  Portal: Menu$1.Portal,
  Backdrop: MenuBackdrop,
  Positioner: MenuPositioner,
  Popup: MenuPopup,
  Arrow: MenuArrow,
  Item: MenuItem,
  Group: MenuGroup,
  GroupLabel: MenuGroupLabel,
  RadioGroup: MenuRadioGroup,
  RadioItem: MenuRadioItem,
  RadioItemIndicator: MenuRadioItemIndicator,
  CheckboxItem: MenuCheckboxItem,
  CheckboxItemIndicator: MenuCheckboxItemIndicator,
  SubmenuTrigger: MenuSubmenuTrigger,
  Separator: MenuSeparator,
  Combobox
};

export { Menu };
