// @base-ui-components/react/cjs/checkbox/indicator/CheckboxIndicatorDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckboxIndicatorDataAttributes = void 0;
let CheckboxIndicatorDataAttributes = exports.CheckboxIndicatorDataAttributes = /*#__PURE__*/function (CheckboxIndicatorDataAttributes) {
  CheckboxIndicatorDataAttributes["checked"] = "data-checked";
  CheckboxIndicatorDataAttributes["unchecked"] = "data-unchecked";
  CheckboxIndicatorDataAttributes["disabled"] = "data-disabled";
  CheckboxIndicatorDataAttributes["readonly"] = "data-readonly";
  CheckboxIndicatorDataAttributes["required"] = "data-required";
  CheckboxIndicatorDataAttributes["startingStyle"] = "data-starting-style";
  CheckboxIndicatorDataAttributes["endingStyle"] = "data-ending-style";
  CheckboxIndicatorDataAttributes["valid"] = "data-valid";
  CheckboxIndicatorDataAttributes["invalid"] = "data-invalid";
  CheckboxIndicatorDataAttributes["touched"] = "data-touched";
  CheckboxIndicatorDataAttributes["dirty"] = "data-dirty";
  CheckboxIndicatorDataAttributes["filled"] = "data-filled";
  CheckboxIndicatorDataAttributes["focused"] = "data-focused";
  return CheckboxIndicatorDataAttributes;
}({});
