// @backstage/plugin-kubernetes-react/dist/components/CustomResources/DefaultCustomResource.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import Accordion from '@material-ui/core/Accordion';
import AccordionDetails from '@material-ui/core/AccordionDetails';
import AccordionSummary from '@material-ui/core/AccordionSummary';
import Grid from '@material-ui/core/Grid';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { DefaultCustomResourceDrawer } from './DefaultCustomResourceDrawer.esm.js';
import { StructuredMetadataTable } from '@backstage/core-components';

const DefaultCustomResourceSummary = ({
  customResource,
  customResourceName
}) => {
  return /* @__PURE__ */ jsx(
    Grid,
    {
      container: true,
      direction: "row",
      justifyContent: "space-between",
      alignItems: "center",
      spacing: 0,
      children: /* @__PURE__ */ jsx(Grid, { xs: 12, item: true, children: /* @__PURE__ */ jsx(
        DefaultCustomResourceDrawer,
        {
          customResource,
          customResourceName
        }
      ) })
    }
  );
};
const DefaultCustomResourceAccordion = ({
  customResource,
  customResourceName,
  defaultExpanded
}) => {
  return /* @__PURE__ */ jsxs(
    Accordion,
    {
      defaultExpanded,
      TransitionProps: { unmountOnExit: true },
      variant: "outlined",
      children: [
        /* @__PURE__ */ jsx(AccordionSummary, { expandIcon: /* @__PURE__ */ jsx(ExpandMoreIcon, {}), children: /* @__PURE__ */ jsx(
          DefaultCustomResourceSummary,
          {
            customResource,
            customResourceName
          }
        ) }),
        /* @__PURE__ */ jsx(AccordionDetails, { children: Object.prototype.hasOwnProperty.call(customResource, "status") && /* @__PURE__ */ jsx(
          StructuredMetadataTable,
          {
            metadata: customResource.status,
            options: { nestedValuesAsYaml: true }
          }
        ) })
      ]
    }
  );
};
const DefaultCustomResourceAccordions = ({
  customResources,
  customResourceName,
  defaultExpanded = false
}) => {
  return /* @__PURE__ */ jsx(
    Grid,
    {
      container: true,
      direction: "column",
      justifyContent: "flex-start",
      alignItems: "flex-start",
      children: customResources.map((cr, i) => /* @__PURE__ */ jsx(Grid, { container: true, item: true, xs: true, children: /* @__PURE__ */ jsx(Grid, { item: true, xs: true, children: /* @__PURE__ */ jsx(
        DefaultCustomResourceAccordion,
        {
          defaultExpanded,
          customResource: cr,
          customResourceName
        }
      ) }) }, i))
    }
  );
};

export { DefaultCustomResourceAccordions };
