// @backstage/canon/dist/components/DataTable/Pagination/DataTablePagination.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { Text } from '../../Text/Text.esm.js';
import { IconButton } from '../../IconButton/IconButton.esm.js';
import clsx from 'clsx';
import { Select } from '../../Select/Select.esm.js';
import { useDataTable } from '../Root/DataTableRoot.esm.js';
import '@remixicon/react';
import { Icon } from '../../Icon/Icon.esm.js';
import '../../Icon/context.esm.js';

const DataTablePagination = forwardRef(
  (props, ref) => {
    const { className, ...rest } = props;
    const { table } = useDataTable();
    const pageIndex = table?.getState().pagination.pageIndex;
    const pageSize = table?.getState().pagination.pageSize;
    const rowCount = table?.getRowCount();
    const fromCount = (pageIndex ?? 0) * (pageSize ?? 10) + 1;
    const toCount = Math.min(
      ((pageIndex ?? 0) + 1) * (pageSize ?? 10),
      rowCount
    );
    return /* @__PURE__ */ jsxs(
      "div",
      {
        ref,
        style: { minWidth: table?.getTotalSize() },
        className: clsx("canon-DataTablePagination", className),
        ...rest,
        children: [
          /* @__PURE__ */ jsx("div", { className: "canon-DataTablePagination--left", children: !table.options.manualPagination && /* @__PURE__ */ jsx(
            Select,
            {
              name: "pageSize",
              size: "small",
              placeholder: "Show 10 results",
              options: [
                { label: "Show 5 results", value: "5" },
                { label: "Show 10 results", value: "10" },
                { label: "Show 20 results", value: "20" },
                { label: "Show 30 results", value: "30" },
                { label: "Show 40 results", value: "40" },
                { label: "Show 50 results", value: "50" }
              ],
              value: pageSize?.toString(),
              onValueChange: (value) => {
                table?.setPageSize(Number(value));
              },
              className: "canon-DataTablePagination--select"
            }
          ) }),
          /* @__PURE__ */ jsxs("div", { className: "canon-DataTablePagination--right", children: [
            /* @__PURE__ */ jsx(Text, { variant: "body", children: `${fromCount} - ${toCount} of ${rowCount}` }),
            /* @__PURE__ */ jsx(
              IconButton,
              {
                variant: "secondary",
                size: "small",
                onClick: () => table?.previousPage(),
                disabled: !table?.getCanPreviousPage(),
                icon: /* @__PURE__ */ jsx(Icon, { name: "chevron-left" })
              }
            ),
            /* @__PURE__ */ jsx(
              IconButton,
              {
                variant: "secondary",
                size: "small",
                onClick: () => table?.nextPage(),
                disabled: !table?.getCanNextPage(),
                icon: /* @__PURE__ */ jsx(Icon, { name: "chevron-right" })
              }
            )
          ] })
        ]
      }
    );
  }
);
DataTablePagination.displayName = "DataTablePagination";

export { DataTablePagination };
