// @base-ui-components/react/cjs/accordion/trigger/AccordionTrigger.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AccordionTrigger = void 0;
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _collapsibleOpenStateMapping = require("../../utils/collapsibleOpenStateMapping");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useEnhancedEffect = require("../../utils/useEnhancedEffect");
var _CollapsibleRootContext = require("../../collapsible/root/CollapsibleRootContext");
var _useCollapsibleTrigger = require("../../collapsible/trigger/useCollapsibleTrigger");
var _AccordionItemContext = require("../item/AccordionItemContext");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * A button that opens and closes the corresponding panel.
 * Renders a `<button>` element.
 *
 * Documentation: [Base UI Accordion](https://base-ui.com/react/components/accordion)
 */

const AccordionTrigger = exports.AccordionTrigger = /*#__PURE__*/React.forwardRef(function AccordionTrigger(props, forwardedRef) {
  const {
    disabled: disabledProp,
    className,
    id,
    render,
    ...otherProps
  } = props;
  const {
    panelId,
    disabled: contextDisabled,
    open,
    setOpen
  } = (0, _CollapsibleRootContext.useCollapsibleRootContext)();
  const {
    getRootProps
  } = (0, _useCollapsibleTrigger.useCollapsibleTrigger)({
    panelId,
    disabled: disabledProp || contextDisabled,
    open,
    rootRef: forwardedRef,
    setOpen
  });
  const {
    state,
    setTriggerId,
    triggerId
  } = (0, _AccordionItemContext.useAccordionItemContext)();
  (0, _useEnhancedEffect.useEnhancedEffect)(() => {
    if (id) {
      setTriggerId(id);
    }
    return () => {
      setTriggerId(undefined);
    };
  }, [id, setTriggerId]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    propGetter: getRootProps,
    render: render ?? 'button',
    state,
    className,
    extraProps: {
      ...otherProps,
      // the `id` prop doesn't go here directly, it updates a context
      // and becomes `triggerId`
      id: triggerId
    },
    customStyleHookMapping: _collapsibleOpenStateMapping.triggerOpenStateMapping
  });
  return renderElement();
});
process.env.NODE_ENV !== "production" ? AccordionTrigger.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: _propTypes.default.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  /**
   * @ignore
   */
  disabled: _propTypes.default.bool,
  /**
   * @ignore
   */
  id: _propTypes.default.string,
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.func])
} : void 0;
