// @backstage/plugin-kubernetes-react/dist/components/CronJobsAccordions/CronJobsDrawer.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { KubernetesStructuredMetadataTableDrawer } from '../KubernetesDrawer/KubernetesStructuredMetadataTableDrawer.esm.js';
import '../KubernetesDrawer/KubernetesDrawer.esm.js';
import '@backstage/core-components';
import '@material-ui/core/FormControlLabel';
import '@material-ui/core/Switch';
import 'js-yaml';
import 'react';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import Chip from '@material-ui/core/Chip';

const CronJobDrawer = ({
  cronJob,
  expanded
}) => {
  const namespace = cronJob.metadata?.namespace;
  return /* @__PURE__ */ jsx(
    KubernetesStructuredMetadataTableDrawer,
    {
      object: cronJob,
      expanded,
      kind: "CronJob",
      renderObject: (cronJobObj) => ({
        schedule: cronJobObj.spec?.schedule ?? "???",
        startingDeadlineSeconds: cronJobObj.spec?.startingDeadlineSeconds ?? "???",
        concurrencyPolicy: cronJobObj.spec?.concurrencyPolicy ?? "???",
        lastScheduleTime: cronJobObj.status?.lastScheduleTime ?? "???"
      }),
      children: /* @__PURE__ */ jsxs(
        Grid,
        {
          container: true,
          direction: "column",
          justifyContent: "flex-start",
          alignItems: "flex-start",
          spacing: 0,
          children: [
            /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Typography, { variant: "body1", children: cronJob.metadata?.name ?? "unknown object" }) }),
            /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Typography, { color: "textSecondary", variant: "subtitle1", children: "CronJob" }) }),
            namespace && /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Chip, { size: "small", label: `namespace: ${namespace}` }) })
          ]
        }
      )
    }
  );
};

export { CronJobDrawer };
