// @backstage/plugin-app/dist/plugins/app/package.json.esm.js
var name = "@backstage/plugin-app";
var version = "0.1.10";
var backstage = {
	role: "frontend-plugin",
	pluginId: "app",
	pluginPackages: [
		"@backstage/plugin-app",
		"@backstage/plugin-app-backend",
		"@backstage/plugin-app-node"
	]
};
var publishConfig = {
	access: "public",
	main: "dist/index.esm.js",
	types: "dist/index.d.ts"
};
var repository = {
	type: "git",
	url: "https://github.com/backstage/backstage",
	directory: "plugins/app"
};
var license = "Apache-2.0";
var sideEffects = false;
var main = "src/index.ts";
var types = "src/index.ts";
var files = [
	"dist"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	start: "backstage-cli package start",
	test: "backstage-cli package test"
};
var dependencies = {
	"@backstage/core-components": "workspace:^",
	"@backstage/core-plugin-api": "workspace:^",
	"@backstage/frontend-plugin-api": "workspace:^",
	"@backstage/integration-react": "workspace:^",
	"@backstage/plugin-permission-react": "workspace:^",
	"@backstage/theme": "workspace:^",
	"@backstage/types": "workspace:^",
	"@material-ui/core": "^4.9.13",
	"@material-ui/icons": "^4.9.1",
	"@material-ui/lab": "^4.0.0-alpha.61",
	"react-use": "^17.2.4"
};
var devDependencies = {
	"@backstage/cli": "workspace:^",
	"@backstage/dev-utils": "workspace:^",
	"@backstage/frontend-test-utils": "workspace:^",
	"@testing-library/jest-dom": "^6.0.0",
	"@testing-library/react": "^16.0.0",
	"@testing-library/user-event": "^14.0.0",
	"@types/react": "^18.0.0",
	msw: "^1.0.0",
	react: "^18.0.2",
	"react-dom": "^18.0.2",
	"react-router-dom": "^6.3.0"
};
var peerDependencies = {
	"@types/react": "^17.0.0 || ^18.0.0",
	react: "^17.0.0 || ^18.0.0",
	"react-dom": "^17.0.0 || ^18.0.0",
	"react-router-dom": "^6.3.0"
};
var peerDependenciesMeta = {
	"@types/react": {
		optional: true
	}
};
var _package = {
	name: name,
	version: version,
	backstage: backstage,
	publishConfig: publishConfig,
	repository: repository,
	license: license,
	sideEffects: sideEffects,
	main: main,
	types: types,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies,
	peerDependencies: peerDependencies,
	peerDependenciesMeta: peerDependenciesMeta
};

export { backstage, _package as default, dependencies, devDependencies, files, license, main, name, peerDependencies, peerDependenciesMeta, publishConfig, repository, scripts, sideEffects, types, version };
