// @backstage/plugin-kubernetes-react/dist/components/DaemonSetsAccordions/DaemonSetsDrawer.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { KubernetesStructuredMetadataTableDrawer } from '../KubernetesDrawer/KubernetesStructuredMetadataTableDrawer.esm.js';
import '../KubernetesDrawer/KubernetesDrawer.esm.js';
import '@backstage/core-components';
import '@material-ui/core/FormControlLabel';
import '@material-ui/core/Switch';
import 'js-yaml';
import 'react';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import Chip from '@material-ui/core/Chip';

const DaemonSetDrawer = ({
  daemonset,
  expanded
}) => {
  const namespace = daemonset.metadata?.namespace;
  return /* @__PURE__ */ jsx(
    KubernetesStructuredMetadataTableDrawer,
    {
      object: daemonset,
      expanded,
      kind: "DaemonSet",
      renderObject: (daemonsetObj) => {
        return {
          updateStrategyType: daemonsetObj.spec?.updateStrategy?.type ?? "???",
          minReadySeconds: daemonsetObj.spec?.minReadySeconds ?? "???",
          revisionHistoryLimit: daemonsetObj.spec?.revisionHistoryLimit ?? "???",
          currentNumberScheduled: daemonsetObj.status?.currentNumberScheduled ?? "???",
          desiredNumberScheduled: daemonsetObj.status?.desiredNumberScheduled ?? "???",
          numberAvailable: daemonsetObj.status?.numberAvailable ?? "???",
          numberMisscheduled: daemonsetObj.status?.numberMisscheduled ?? "???",
          numberReady: daemonsetObj.status?.numberReady ?? "???"
        };
      },
      children: /* @__PURE__ */ jsxs(
        Grid,
        {
          container: true,
          direction: "column",
          justifyContent: "flex-start",
          alignItems: "flex-start",
          spacing: 0,
          children: [
            /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Typography, { variant: "body1", children: daemonset.metadata?.name ?? "unknown object" }) }),
            /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Typography, { color: "textSecondary", variant: "subtitle1", children: "DaemonSet" }) }),
            namespace && /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Chip, { size: "small", label: `namespace: ${namespace}` }) })
          ]
        }
      )
    }
  );
};

export { DaemonSetDrawer };
