// @backstage/plugin-kubernetes-react/dist/components/JobsAccordions/JobsDrawer.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { KubernetesStructuredMetadataTableDrawer } from '../KubernetesDrawer/KubernetesStructuredMetadataTableDrawer.esm.js';
import '../KubernetesDrawer/KubernetesDrawer.esm.js';
import '@backstage/core-components';
import '@material-ui/core/FormControlLabel';
import '@material-ui/core/Switch';
import 'js-yaml';
import 'react';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';

const JobDrawer = ({
  job,
  expanded
}) => {
  return /* @__PURE__ */ jsx(
    KubernetesStructuredMetadataTableDrawer,
    {
      object: job,
      expanded,
      kind: "Job",
      renderObject: (jobObj) => {
        return {
          parallelism: jobObj.spec?.parallelism ?? "???",
          completions: jobObj.spec?.completions ?? "???",
          backoffLimit: jobObj.spec?.backoffLimit ?? "???",
          startTime: jobObj.status?.startTime ?? "???"
        };
      },
      children: /* @__PURE__ */ jsxs(
        Grid,
        {
          container: true,
          direction: "column",
          justifyContent: "flex-start",
          alignItems: "flex-start",
          spacing: 0,
          children: [
            /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Typography, { variant: "body1", children: job.metadata?.name ?? "unknown object" }) }),
            /* @__PURE__ */ jsx(Grid, { item: true, children: /* @__PURE__ */ jsx(Typography, { color: "textSecondary", variant: "subtitle1", children: "Job" }) })
          ]
        }
      )
    }
  );
};

export { JobDrawer };
