// @base-ui-components/react/cjs/utils/defaultRenderFunctions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultRenderFunctions = void 0;
var React = _interopRequireWildcard(require("react"));
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const defaultRenderFunctions = exports.defaultRenderFunctions = {
  button: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
      type: "button",
      ...props
    });
  },
  div: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      ...props
    });
  },
  h2: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("h2", {
      ...props
    });
  },
  h3: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("h3", {
      ...props
    });
  },
  output: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("output", {
      ...props
    });
  },
  p: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("p", {
      ...props
    });
  },
  span: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      ...props
    });
  },
  a: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("a", {
      ...props
    });
  },
  label: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("label", {
      ...props
    });
  },
  input: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
      ...props
    });
  },
  fieldset: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("fieldset", {
      ...props
    });
  },
  form: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("form", {
      ...props
    });
  },
  img: props => {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
      alt: "",
      ...props
    });
  }
};
