// @base-ui-components/react/esm/popover/title/PopoverTitle.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
import { usePopoverRootContext } from '../root/PopoverRootContext.js';
import { usePopoverTitle } from './usePopoverTitle.js';
const state = {};

/**
 * A heading that labels the popover.
 * Renders an `<h2>` element.
 *
 * Documentation: [Base UI Popover](https://base-ui.com/react/components/popover)
 */
const PopoverTitle = /*#__PURE__*/React.forwardRef(function PopoverTitle(props, forwardedRef) {
  const {
    render,
    className,
    ...otherProps
  } = props;
  const {
    setTitleId
  } = usePopoverRootContext();
  const {
    getTitleProps
  } = usePopoverTitle({
    titleId: otherProps.id,
    setTitleId
  });
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getTitleProps,
    render: render ?? 'h2',
    className,
    state,
    ref: forwardedRef,
    extraProps: otherProps
  });
  return renderElement();
});
process.env.NODE_ENV !== "production" ? PopoverTitle.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * @ignore
   */
  id: PropTypes.string,
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
export { PopoverTitle };
