// @base-ui-components/react/esm/progress/value/ProgressValue.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { mergeProps } from '../../merge-props/index.js';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
import { useProgressRootContext } from '../root/ProgressRootContext.js';
import { progressStyleHookMapping } from '../root/styleHooks.js';
/**
 * A text label displaying the current value.
 * Renders a `<span>` element.
 *
 * Documentation: [Base UI Progress](https://base-ui.com/react/components/progress)
 */
const ProgressValue = /*#__PURE__*/React.forwardRef(function ProgressValue(props, forwardedRef) {
  const {
    className,
    render,
    children,
    ...otherProps
  } = props;
  const {
    value,
    formattedValue,
    state
  } = useProgressRootContext();
  const getValueProps = React.useCallback((externalProps = {}) => {
    const formattedValueArg = value == null ? 'indeterminate' : formattedValue;
    const formattedValueDisplay = value == null ? null : formattedValue;
    return mergeProps({
      'aria-hidden': true,
      children: typeof children === 'function' ? children(formattedValueArg, value) : formattedValueDisplay
    }, externalProps);
  }, [children, value, formattedValue]);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getValueProps,
    render: render ?? 'span',
    className,
    state,
    ref: forwardedRef,
    extraProps: otherProps,
    customStyleHookMapping: progressStyleHookMapping
  });
  return renderElement();
});
process.env.NODE_ENV !== "production" ? ProgressValue.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.func,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
export { ProgressValue };
