// @base-ui-components/react/esm/accordion/header/AccordionHeader.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
import { useAccordionItemContext } from '../item/AccordionItemContext.js';
import { accordionStyleHookMapping } from '../item/styleHooks.js';

/**
 * A heading that labels the corresponding panel.
 * Renders an `<h3>` element.
 *
 * Documentation: [Base UI Accordion](https://base-ui.com/react/components/accordion)
 */
const AccordionHeader = /*#__PURE__*/React.forwardRef(function AccordionHeader(props, forwardedRef) {
  const {
    render,
    className,
    ...other
  } = props;
  const {
    state
  } = useAccordionItemContext();
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'h3',
    state,
    className,
    ref: forwardedRef,
    extraProps: other,
    customStyleHookMapping: accordionStyleHookMapping
  });
  return renderElement();
});
export { AccordionHeader };
process.env.NODE_ENV !== "production" ? AccordionHeader.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
