// @base-ui-components/react/esm/slider/indicator/SliderIndicator.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
import { useSliderRootContext } from '../root/SliderRootContext.js';
import { sliderStyleHookMapping } from '../root/styleHooks.js';
import { useSliderIndicator } from './useSliderIndicator.js';

/**
 * Visualizes the current value of the slider.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Slider](https://base-ui.com/react/components/slider)
 */
const SliderIndicator = /*#__PURE__*/React.forwardRef(function SliderIndicator(props, forwardedRef) {
  const {
    render,
    className,
    ...otherProps
  } = props;
  const {
    disabled,
    orientation,
    state,
    percentageValues
  } = useSliderRootContext();
  const {
    getRootProps
  } = useSliderIndicator({
    disabled,
    orientation,
    percentageValues
  });
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getRootProps,
    render: render ?? 'div',
    state,
    className,
    ref: forwardedRef,
    extraProps: otherProps,
    customStyleHookMapping: sliderStyleHookMapping
  });
  return renderElement();
});
export { SliderIndicator };
process.env.NODE_ENV !== "production" ? SliderIndicator.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
