// @backstage/plugin-kubernetes/dist/package.json.esm.js
var name = "@backstage/plugin-kubernetes";
var version = "0.12.8";
var description = "A Backstage plugin that integrates towards Kubernetes";
var backstage = {
	role: "frontend-plugin",
	pluginId: "kubernetes",
	pluginPackages: [
		"@backstage/plugin-kubernetes",
		"@backstage/plugin-kubernetes-backend",
		"@backstage/plugin-kubernetes-common",
		"@backstage/plugin-kubernetes-node",
		"@backstage/plugin-kubernetes-react"
	]
};
var publishConfig = {
	access: "public"
};
var keywords = [
	"backstage",
	"kubernetes"
];
var homepage = "https://backstage.io";
var repository = {
	type: "git",
	url: "https://github.com/backstage/backstage",
	directory: "plugins/kubernetes"
};
var license = "Apache-2.0";
var sideEffects = false;
var exports = {
	".": "./src/index.ts",
	"./alpha": "./src/alpha.ts",
	"./package.json": "./package.json"
};
var main = "src/index.ts";
var types = "src/index.ts";
var typesVersions = {
	"*": {
		alpha: [
			"src/alpha.ts"
		],
		"package.json": [
			"package.json"
		]
	}
};
var files = [
	"dist"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	start: "backstage-cli package start",
	test: "backstage-cli package test"
};
var dependencies = {
	"@backstage/catalog-model": "workspace:^",
	"@backstage/core-compat-api": "workspace:^",
	"@backstage/core-components": "workspace:^",
	"@backstage/core-plugin-api": "workspace:^",
	"@backstage/frontend-plugin-api": "workspace:^",
	"@backstage/plugin-catalog-react": "workspace:^",
	"@backstage/plugin-kubernetes-common": "workspace:^",
	"@backstage/plugin-kubernetes-react": "workspace:^",
	"@backstage/plugin-permission-react": "workspace:^",
	"@kubernetes-models/apimachinery": "^2.0.0",
	"@kubernetes-models/base": "^5.0.0",
	"@kubernetes/client-node": "1.1.2",
	"@material-ui/core": "^4.12.2",
	"@xterm/addon-attach": "^0.11.0",
	"@xterm/addon-fit": "^0.10.0",
	"@xterm/xterm": "^5.5.0",
	cronstrue: "^2.2.0",
	"js-yaml": "^4.0.0",
	"kubernetes-models": "^4.1.0",
	lodash: "^4.17.21",
	luxon: "^3.0.0"
};
var devDependencies = {
	"@backstage/cli": "workspace:^",
	"@backstage/dev-utils": "workspace:^",
	"@backstage/test-utils": "workspace:^",
	"@testing-library/dom": "^10.0.0",
	"@testing-library/jest-dom": "^6.0.0",
	"@testing-library/react": "^16.0.0",
	"@types/react": "^18.0.0",
	react: "^18.0.2",
	"react-dom": "^18.0.2",
	"react-router-dom": "^6.3.0"
};
var peerDependencies = {
	"@types/react": "^17.0.0 || ^18.0.0",
	react: "^17.0.0 || ^18.0.0",
	"react-dom": "^17.0.0 || ^18.0.0",
	"react-router-dom": "^6.3.0"
};
var peerDependenciesMeta = {
	"@types/react": {
		optional: true
	}
};
var _package = {
	name: name,
	version: version,
	description: description,
	backstage: backstage,
	publishConfig: publishConfig,
	keywords: keywords,
	homepage: homepage,
	repository: repository,
	license: license,
	sideEffects: sideEffects,
	exports: exports,
	main: main,
	types: types,
	typesVersions: typesVersions,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies,
	peerDependencies: peerDependencies,
	peerDependenciesMeta: peerDependenciesMeta
};

export { backstage, _package as default, dependencies, description, devDependencies, exports, files, homepage, keywords, license, main, name, peerDependencies, peerDependenciesMeta, publishConfig, repository, scripts, sideEffects, types, typesVersions, version };
