// @backstage/plugin-kubernetes-react/dist/components/Pods/Events/Events.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import Avatar from '@material-ui/core/Avatar';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemAvatar from '@material-ui/core/ListItemAvatar';
import ListItemText from '@material-ui/core/ListItemText';
import Tooltip from '@material-ui/core/Tooltip';
import Typography from '@material-ui/core/Typography';
import InfoIcon from '@material-ui/icons/Info';
import WarningIcon from '@material-ui/icons/Warning';
import { DateTime } from 'luxon';
import { useEvents } from './useEvents.esm.js';
import Skeleton from '@material-ui/lab/Skeleton';
import { DismissableBanner } from '@backstage/core-components';

const getAvatarByType = (type) => {
  return /* @__PURE__ */ jsx(ListItemAvatar, { children: /* @__PURE__ */ jsx(Avatar, { children: type === "Warning" ? /* @__PURE__ */ jsx(WarningIcon, {}) : /* @__PURE__ */ jsx(InfoIcon, {}) }) });
};
const EventsContent = ({
  events,
  warningEventsOnly
}) => {
  if (events.length === 0) {
    return /* @__PURE__ */ jsx(Typography, { children: "No events found" });
  }
  return /* @__PURE__ */ jsx(Container, { children: /* @__PURE__ */ jsx(Grid, { children: /* @__PURE__ */ jsx(List, { children: events.filter((event) => {
    if (warningEventsOnly) {
      return event.type === "Warning";
    }
    return true;
  }).map((event, index) => {
    const timeAgo = event.metadata.creationTimestamp ? DateTime.fromISO(event.metadata.creationTimestamp).toRelative(
      {
        locale: "en"
      }
    ) : "unknown";
    return /* @__PURE__ */ jsxs(ListItem, { children: [
      /* @__PURE__ */ jsx(Tooltip, { title: `${event.type ?? ""} event`, children: getAvatarByType(event.type) }),
      /* @__PURE__ */ jsx(
        ListItemText,
        {
          primary: `First event ${timeAgo} (count: ${event.count})`,
          secondary: `${event.reason}: ${event.message}`
        }
      )
    ] }, `${event.metadata.name}-${index}`);
  }) }) }) });
};
const Events = ({
  involvedObjectName,
  namespace,
  clusterName,
  warningEventsOnly
}) => {
  const { value, error, loading } = useEvents({
    involvedObjectName,
    namespace,
    clusterName
  });
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    error && /* @__PURE__ */ jsx(
      DismissableBanner,
      {
        ...{
          message: error.message,
          variant: "error",
          fixed: false
        },
        id: "events"
      }
    ),
    loading && /* @__PURE__ */ jsx(Skeleton, { variant: "rect", width: "100%", height: "100%" }),
    !loading && value !== void 0 && /* @__PURE__ */ jsx(EventsContent, { warningEventsOnly, events: value })
  ] });
};

export { Events, EventsContent };
