// @backstage/canon/dist/components/Table/Table.esm.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import clsx from 'clsx';
import { TableCell } from './TableCell/TableCell.esm.js';
import { TableCellText } from './TableCellText/TableCellText.esm.js';
import { TableCellLink } from './TableCellLink/TableCellLink.esm.js';
import { TableCellProfile } from './TableCellProfile/TableCellProfile.esm.js';

const TableRoot = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx("table", { ref, className: clsx("canon-TableRoot", className), ...props }));
TableRoot.displayName = "TableRoot";
const TableHeader = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  "thead",
  {
    ref,
    className: clsx("canon-TableHeader", className),
    ...props
  }
));
TableHeader.displayName = "TableHeader";
const TableBody = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx("tbody", { ref, className: clsx("canon-TableBody", className), ...props }));
TableBody.displayName = "TableBody";
const TableRow = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx("tr", { ref, className: clsx("canon-TableRow", className), ...props, children: props.children }));
TableRow.displayName = "TableRow";
const TableHead = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx("th", { ref, className: clsx("canon-TableHead", className), ...props }));
TableHead.displayName = "TableHead";
const TableCaption = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  "caption",
  {
    ref,
    className: clsx("canon-TableCaption", className),
    ...props
  }
));
TableCaption.displayName = "TableCaption";
const Table = {
  Root: TableRoot,
  Header: TableHeader,
  Body: TableBody,
  Head: TableHead,
  Row: TableRow,
  Cell: TableCell,
  CellText: TableCellText,
  CellLink: TableCellLink,
  CellProfile: TableCellProfile,
  Caption: TableCaption
};

export { Table };
